File: parser/src/main/java/com/aaroncoplan/waterfall/parser/parsing/SyntaxErrorListener.java
package com.aaroncoplan.waterfall.parser.parsing;

import java.util.ArrayList;
import java.util.List;

import org.antlr.v4.runtime.BaseErrorListener;
import org.antlr.v4.runtime.RecognitionException;
import org.antlr.v4.runtime.Recognizer;
import org.antlr.v4.runtime.misc.ParseCancellationException;

class SyntaxErrorListener extends BaseErrorListener {
    private final String fileName;
    private final List<String> syntaxErrors;

    public SyntaxErrorListener(String fileName) {
        this.fileName = fileName;
        this.syntaxErrors = new ArrayList<>();
    }

    public List<String> getSyntaxErrors() {
        return syntaxErrors;
    }

    @Override
    public void syntaxError(
        Recognizer<?, ?> recognizer,
        Object offendingSymbol,
        int line,
        int charPositionInLine,
        String msg,
        RecognitionException e
    ) throws ParseCancellationException {
        final String errorMessage = fileName + " line " + line + ":" + charPositionInLine + " " + msg;
        syntaxErrors.add(errorMessage);
    }

}

File: parser/src/main/kotlin/com/aaroncoplan/waterfall/parser/SyntaxErrorListener.kt
package com.aaroncoplan.waterfall.parser

import org.antlr.v4.runtime.BaseErrorListener
import org.antlr.v4.runtime.RecognitionException
import org.antlr.v4.runtime.Recognizer
import org.antlr.v4.runtime.misc.ParseCancellationException

internal class SyntaxErrorListener(private val fileName: String) : BaseErrorListener() {

    private val syntaxErrors: MutableList<String> = mutableListOf()

    fun getSyntaxErrors(): List<String> = syntaxErrors

    @Throws(ParseCancellationException::class)
    override fun syntaxError(
        recognizer: Recognizer<*, *>?,
        offendingSymbol: Any?,
        line: Int,
        charPositionInLine: Int,
        msg: String?,
        e: RecognitionException?
    ) {
        val errorMessage = "$fileName line $line:$charPositionInLine $msg"
        syntaxErrors.add(errorMessage)
    }
}
